### PR TITLE
add generalised dice loss and test, updated dice loss

### LINKED ISSUE #116: Port generalized Dice loss
a subtask of #84, port a generalized dice loss function into Monai

### PR DESCRIPTION
**Pull request by wyli**
Fixes #116.

### Description
added generalised dice loss and tests.
updated dice loss

### Status
**Ready**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### CODE REVIEW

File: monai/losses/dice.py [hunk 1]
```diff
@@ -69,12 +69,93 @@ def forward(self, pred, ground, smooth=1e-5):
         assert tsum.shape == psum.shape, ("Ground truth one-hot has differing shape (%r) from source (%r)" %
                                           (tsum.shape, psum.shape))
 
-        batchsize = ground.size(0)
-        tsum = tsum.float().view(batchsize, -1)
-        psum = psum.view(batchsize, -1)
+        batchsize, n_classes = tsum.shape[:2]
+        tsum = tsum.float().view(batchsize, n_classes, -1)
+        psum = psum.view(batchsize, n_classes, -1)
```
Comment (wyli) [inline monai/losses/dice.py]: previously the dice scores were not averaged over class channels

File: monai/losses/dice.py [hunk 2]
```diff
@@ -69,12 +69,93 @@ def forward(self, pred, ground, smooth=1e-5):
...
+        if self.do_sigmoid:
+            psum = psum.sigmoid()  # use sigmoid activation
+        if pred.shape[1] == 1:
+            if self.do_softmax:
+                raise ValueError('do_softmax is not compatible with single channel prediction.')
+            if not self.include_background:
+                raise RuntimeWarning('single channel prediction, `include_background=False` ignored.')
+            tsum = ground
+        else:  # multiclass dice loss
+            if self.do_softmax:
```
Comment (ericspod) [inline monai/losses/dice.py]: If it's the case that do_softmax and so_sigmoid are never both true, and members don't change throughout the object's lifetime, this check should be the constructor.

File: monai/losses/dice.py [hunk 3]
```diff
@@ -69,12 +69,93 @@ def forward(self, pred, ground, smooth=1e-5):
...
+            raise ValueError("Ground truth should have only a single channel, shape is " + str(ground.shape))
+
+        psum = pred.float()
+        if self.do_sigmoid:
+            psum = psum.sigmoid()  # use sigmoid activation
+        if pred.shape[1] == 1:
+            if self.do_softmax:
+                raise ValueError('do_softmax is not compatible with single channel prediction.')
+            if not self.include_background:
+                raise RuntimeWarning('single channel prediction, `include_background=False` ignored.')
```
Comment (ericspod) [inline monai/losses/dice.py]: Won't this raise an exception? We should maybe use the `warnings` library.

File: monai/losses/dice.py [hunk 4]
```diff
@@ -69,12 +69,93 @@ def forward(self, pred, ground, smooth=1e-5):
...
+            do_sigmoid (bool): If True, apply a sigmoid function to the prediction.
+            do_softmax (bool): If True, apply a softmax function to the prediction.
+            w_type ('square'|'simple'|'uniform'): type of function to transform ground truth volume to a weight factor.
+        """
+        super().__init__()
+        self.include_background = include_background
+        self.do_sigmoid = do_sigmoid
+        self.do_softmax = do_softmax
+
+        self.w_func = lambda x: torch.ones_like(x)
```
Comment (ericspod) [inline monai/losses/dice.py]: It's perhaps stylistic but one could just have `self.w_func = torch.ones_like`.

File: monai/losses/dice.py [hunk 5]
```diff
@@ -69,12 +69,93 @@ def forward(self, pred, ground, smooth=1e-5):
...
+        batchsize, n_classes = tsum.shape[:2]
+        tsum = tsum.float().view(batchsize, n_classes, -1)
+        psum = psum.view(batchsize, n_classes, -1)
 
         intersection = psum * tsum
         sums = psum + tsum
 
-        score = 2.0 * (intersection.sum(1) + smooth) / (sums.sum(1) + smooth)
-        return 1 - score.sum() / batchsize
+        score = 2.0 * (intersection.sum(2) + smooth) / (sums.sum(2) + smooth)
```
Comment (ericspod) [inline monai/losses/dice.py]: Should be `score = 2.0 * intersection.sum(2) / (sums.sum(2) + smooth)` if we're removing the smoothing value from the numerator.

Comment (wyli) [reply monai/losses/dice.py]: this ticket says no... https://github.com/NifTK/NiftyNet/issues/249
